### ISSUE TITLE
Any wait to avoid sending headers while generating more than one page?

### ISSUE THREAD

**Issue by fvsch** [getkirby/kirby#446]
Hello, 
I need this for the StaticBuilder plugin, but generally speaking it’s useful to be able to generate the response body for a given page (be it plain text, HTML, JSON or something else), and possibly for several pages, but *not* send the HTTP headers.

Right now I’m using `$body = $kirby->render($page, [], false);` which seems to work fine for generating the right content. The controller is called, then the template, everything is fine.

Problems arise when I have code in the controller or template that sets some HTTP headers. They affect the main response, which is not something I want. If I generate a HTML page report on how the script just generated some pages or rebuilt their cache or something, I don't want to be redirected or receiving a `Content-Type: application/atom+xml`, obviously.

I hear it’s a common PHP problem, and perhaps not something Kirby can do anything about. But if you have some pointers on how Kirby might help, or on some possible PHP solutions, that would be great. Also if there is a generic PHP solution, maybe it could be provided as a utility in Kirby in a future version?

It would be nice if Kirby provided a way to store headers then send them later (and conditionally), and generate a page and return an array with its intended headers + response body (which would help when trying to decide what extension to use for a static file generated from a Kirby page).

**Comment by bastianallgeier** [comment]
Did you try http://php.net/manual/en/function.header-remove.php already?

**Comment by fvsch** [comment]
Yep, it seemed to have no effect. :/
Maybe my tests were wrong.

To narrow down this issue a bit: since Kirby has the ability to send HTTP headers only when calling the render method with `$headers=true`, I was wondering how that was managed and if developers could piggy-back on this feature.

Like setting your headers with `$page->setHeader('Something', 'xyz');` and letting Kirby do the actual sending (or not). Plus the ability to retrieve headers set this way (and set by Kirby). Is that already possible? If not, would it be doable?

**Comment by fvsch** [comment]
Update:

- `header_remove` didn't work when used just after calling `$kirby->render(…)`
- but it does seem to work when used in my routes.

I can probably close https://github.com/fvsch/kirby-staticbuilder/issues/7, but of course if there's a way to improve user code so my plugin and others don't have to use the `header_remove` nuclear bomb, that could be great.

**Comment by lukasbestle** [comment]
The `$headers` option for the `kirby()->render()` method is to disable sending the "template specific headers". I don't know if that's even documented somewhere, but it basically works like this:

```php
c::set('headers', array(
  'sitemap' => function($page) {
    header('Content-Type: application/atom+xml');
  }
));
```

**Comment by fvsch** [comment]
Looks like a good option!

> Edit: The option is documented in the docs.

Does it accept a $page object like in your example?
If yes, this is missing from the docs.

**Comment by fvsch** [comment]
So I got `header_remove()` to work, which fixes the issue in my plugin.

Also on a project I moved to using the `headers` option to change the content-type of some templates. Other templates have redirections (with the `go()` helper), and I’m not sure it would be possible to have them in the `headers` config (they kind of depend on what happens in the controller).

I’m closing this issue, it’s mostly answered and fixed for me.
